allennlp/modules/elmo.py

```diff
@@ -681,3 +685,148 @@ def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+    """
+    Given a word :class:`Vocabulary` and the weights for the softmax
+    layer, return the log probabilities from the ELMo model. The `forward()`
+    method requires you to run :class:`_ElmoBiLm` and `batch_to_ids`
+    beforehand to calculate the probabilities.
+
+    Parameters
+    ----------
+    softmax_weight_file : ``str``, required
+        The .hdf5 file for the softmax weights.
+    softmax_vocab_file : ``str``, optional, (default None)
+        The .txt file for the vocabulary used for training ELMo.
+    chunk_size : ``int``, optional, (default 256)
+        The chunk size of the softmax layer. The default (256) takes about
+        10GB of memory!
+    """
+
+    def __init__(self,
+                 softmax_weight_file: str,
+                 softmax_vocab_file: str = None,
+                 chunk_size: int = 256):
+
+        super(_ElmoSoftmax, self).__init__()
+
+        self.softmax_weight_file = softmax_weight_file
+        self.softmax_vocab_file = softmax_vocab_file
+        self.chunk_size = chunk_size
+
+        if self.softmax_vocab_file:
+            self.vocab = self._load_vocab(self.softmax_vocab_file)
+        else:
+            self.vocab = None
+        self.fc_layer = self._load_softmax_weights(self.softmax_weight_file)
+        self.vocab_size = self.fc_layer.out_features
+        self.hidden_size = self.fc_layer.in_features
+
+    @staticmethod
+    def _load_vocab(softmax_vocab_file: str) -> Vocabulary:
+        vocab = Vocabulary()
+        vocab._oov_token = '<UNK>'
+        vocab.set_from_file(cached_path(softmax_vocab_file), is_padded=False)
+        return vocab
+
+    @staticmethod
+    def _load_softmax_weights(
+            softmax_weight_file: str) -> torch.nn.Linear:
+
+        with h5py.File(cached_path(softmax_weight_file), 'r') as fin:
+            W = numpy.array(fin['softmax']['W'])
+            b = numpy.array(fin['softmax']['b'])
+
+        vocab_size = W.shape[0]
+        hidden_size = W.shape[1]
+        fc_layer = torch.nn.Linear(hidden_size, vocab_size)
+
+        fc_layer.weight.data.copy_(
+            torch.FloatTensor(W))
+        fc_layer.bias.data.copy_(
+            torch.FloatTensor(b))
+
+        return fc_layer
+
+    def _chunked_log_probs(self,
+                           activation: torch.Tensor) -> torch.Tensor:
+        """
+        Do the softmax in chunks so the gpu ram doesn't explode.
+
+        Parameters
+        ----------
+        activation : ``torch.Tensor``, required
+            Results of the ELMo embedder.
+
+        Returns
+        -------
+        ``torch.Tensor`` with all of the log probabilities.
+        """
+        all_logprobs = []
+        num_chunks = (activation.size(0) - 1) // self.chunk_size + 1
+        for activation_chunk in torch.chunk(activation, num_chunks, dim=0):
+            chunk_size, sequence_length = activation_chunk.size()[:2]
+
+            time_indexer = (torch.arange(0, chunk_size * sequence_length)
```

Comment:
What is the purpose of these indexers?

Seems the only thing needed inside this for loop is
```all_log_probs.append(torch.nn.functional.log_softmax(self.fc_layer(activation_chunk), dim=-1))```

??